File: intrinsic/util/proto/descriptors.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/util/proto/descriptors.h"

#include <queue>
#include <string>

#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "absl/status/status.h"
#include "absl/strings/str_format.h"
#include "google/protobuf/descriptor.h"
#include "google/protobuf/descriptor.pb.h"
#include "google/protobuf/descriptor_database.h"
#include "intrinsic/util/status/status_macros.h"

namespace intrinsic {

namespace {

void MergeFileDescriptorSetImpl(const google::protobuf::Descriptor& descriptor,
                                google::protobuf::FileDescriptorSet& set,
                                bool include_source_code_info) {
  absl::flat_hash_set<std::string> visited;
  for (const auto& file : set.file()) {
    visited.emplace(file.name());
  }
  std::queue<const google::protobuf::FileDescriptor*> queue;
  queue.push(descriptor.file());
  while (!queue.empty()) {
    const google::protobuf::FileDescriptor* current = queue.front();
    queue.pop();
    if (visited.contains(current->name())) {
      // There's already a file descriptor with that name, so assume the
      // dependencies will be there too.
      continue;
    }
    visited.emplace(current->name());
    google::protobuf::FileDescriptorProto* proto = set.add_file();
    current->CopyTo(proto);
    if (include_source_code_info) {
      current->CopySourceCodeInfoTo(proto);
    }
    for (int i = 0; i < current->dependency_count(); i++) {
      queue.push(current->dependency(i));
    }
  }
}

void MergeFileDescriptorSetsImpl(
    const google::protobuf::FileDescriptorSet& src_set,
    google::protobuf::FileDescriptorSet& dst_set,
    bool include_source_code_info) {
  absl::flat_hash_set<std::string> dst_files;
  for (const auto& dst_file : dst_set.file()) {
    dst_files.emplace(dst_file.name());
  }
  for (const google::protobuf::FileDescriptorProto& src_file : src_set.file()) {
    if (dst_files.contains(src_file.name())) {
      continue;
    }
    dst_files.insert(src_file.name());
    google::protobuf::FileDescriptorProto* proto = dst_set.add_file();
    *proto = src_file;
    if (!include_source_code_info) {
      proto->clear_source_code_info();
    }
  }
}

}  // namespace

google::protobuf::FileDescriptorSet GenFileDescriptorSet(
    const google::protobuf::Descriptor& descriptor) {
  google::protobuf::FileDescriptorSet out;
  MergeFileDescriptorSetImpl(descriptor, out,
                             /*include_source_code_info=*/false);
  return out;
}

google::protobuf::FileDescriptorSet GenFileDescriptorSetWithSourceInfo(
    const google::protobuf::Descriptor& descriptor) {
  google::protobuf::FileDescriptorSet out;
  MergeFileDescriptorSetImpl(descriptor, out,
                             /*include_source_code_info=*/true);
  return out;
}

void MergeFileDescriptorSet(const google::protobuf::Descriptor& descriptor,
                            google::protobuf::FileDescriptorSet& set) {
  MergeFileDescriptorSetImpl(descriptor, set,
                             /*include_source_code_info=*/false);
}

void MergeFileDescriptorSetWithSourceInfo(
    const google::protobuf::Descriptor& descriptor,
    google::protobuf::FileDescriptorSet& set) {
  MergeFileDescriptorSetImpl(descriptor, set,
                             /*include_source_code_info=*/true);
}

void MergeFileDescriptorSet(const google::protobuf::FileDescriptorSet& src_set,
                            google::protobuf::FileDescriptorSet& dst_set) {
  MergeFileDescriptorSetsImpl(src_set, dst_set,
                              /*include_source_code_info=*/false);
}

void MergeFileDescriptorSetWithSourceInfo(
    const google::protobuf::FileDescriptorSet& src_set,
    google::protobuf::FileDescriptorSet& dst_set) {
  MergeFileDescriptorSetsImpl(src_set, dst_set,
                              /*include_source_code_info=*/true);
}

absl::Status AddToDescriptorDatabase(
    google::protobuf::SimpleDescriptorDatabase* db,
    const google::protobuf::FileDescriptorProto& file_descriptor,
    absl::flat_hash_map<std::string, google::protobuf::FileDescriptorProto>&
        file_by_name) {
  for (const std::string& dependency : file_descriptor.dependency()) {
    if (auto fd_iter = file_by_name.find(dependency);
        fd_iter != file_by_name.end()) {
      // add dependency now and remove from elements to visit
      google::protobuf::FileDescriptorProto dependency_file_descriptor =
          fd_iter->second;
      file_by_name.erase(fd_iter);
      INTR_RETURN_IF_ERROR(AddToDescriptorDatabase(
          db, dependency_file_descriptor, file_by_name));
    }
  }
  if (!db->Add(file_descriptor)) {
    return absl::InvalidArgumentError(
        absl::StrFormat("Failed to add descriptor '%s' to descriptor database",
                        file_descriptor.name()));
  }
  return absl::OkStatus();
}

absl::Status PopulateDescriptorDatabase(
    google::protobuf::SimpleDescriptorDatabase* db,
    const google::protobuf::FileDescriptorSet& file_descriptor_set) {
  absl::flat_hash_map<std::string, google::protobuf::FileDescriptorProto>
      file_by_name;
  for (const google::protobuf::FileDescriptorProto& file_descriptor :
       file_descriptor_set.file()) {
    file_by_name[file_descriptor.name()] = file_descriptor;
  }
  while (!file_by_name.empty()) {
    google::protobuf::FileDescriptorProto file_descriptor =
        file_by_name.begin()->second;
    file_by_name.erase(file_by_name.begin());
    INTR_RETURN_IF_ERROR(
        AddToDescriptorDatabase(db, file_descriptor, file_by_name));
  }
  return absl::OkStatus();
}

}  // namespace intrinsic

File: intrinsic/util/proto/descriptors.py
# Copyright 2023 Intrinsic Innovation LLC

"""Helpers for viewing and manipulating proto descriptors."""

from collections.abc import Iterable
import textwrap

from google.protobuf import descriptor
from google.protobuf import descriptor_pb2
from google.protobuf import descriptor_pool


def _add_file_and_imports(
    file_descriptors: dict[str, descriptor.FileDescriptor],
    current_file: descriptor.FileDescriptor,
):
  """Adds the current file to file_descriptors if not present.

  Recursively adds dependencies until all dependencies (along with transients)
  are present in file_descriptors.

  Args:
    file_descriptors: A dictionary mapping file descriptor name to file
      descriptors. Modified to add transitive dependencies of current_file
    current_file: The file descriptor of the current file.
  """
  if current_file.name in file_descriptors:
    return

  file_descriptors[current_file.name] = current_file
  for dependency_file_descriptor in current_file.dependencies:
    _add_file_and_imports(file_descriptors, dependency_file_descriptor)


def gen_file_descriptor_set(
    msg_descriptor: descriptor.Descriptor | Iterable[descriptor.Descriptor],
) -> descriptor_pb2.FileDescriptorSet:
  """Generates a FileDescriptorSet given a Descriptor.

  Args:
    msg_descriptor: The message descriptor(s) of interest.

  Returns:
    A descriptor_pb2.FileDescriptorSet containing the file descriptors of all
    transitive dependencies of msg_descriptor.
  """
  msg_descriptors: Iterable[descriptor.Descriptor] = (
      msg_descriptor
      if isinstance(msg_descriptor, Iterable)
      else [msg_descriptor]
  )
  file_descriptor_set = descriptor_pb2.FileDescriptorSet()
  file_descriptors: dict[str, descriptor.FileDescriptor] = {}
  for msg_descriptor in msg_descriptors:
    _add_file_and_imports(file_descriptors, msg_descriptor.file)
    for file_descriptor in file_descriptors.values():
      file_descriptor.CopyToProto(file_descriptor_set.file.add())
  return file_descriptor_set


def merge_file_descriptor_sets(
    sets: Iterable[descriptor_pb2.FileDescriptorSet],
) -> descriptor_pb2.FileDescriptorSet:
  """Merges the files of the given file descriptor sets into a single set.

  Merges the given file descriptor sets into a single set without duplicate
  files. If a file name appears in more than one input file descriptor set this
  checks that all instances are identical.

  Args:
    sets: The file descriptor sets to merge.

  Returns:
    The merged file descriptor set.

  Raises:
    TypeError: If a file is contained in more that one input file descriptor set
      but the file's instances are not identical.
  """
  files: dict[str, descriptor_pb2.FileDescriptorProto] = {}
  result_set = descriptor_pb2.FileDescriptorSet()

  for input_set in sets:
    for file in input_set.file:
      if file.name not in files:
        # This makes a copy of 'file'
        result_set.file.append(file)
        files[file.name] = file
      elif file != files[file.name]:
        raise ValueError(
            f'"{file.name}" contained in more than one input file descriptor'
            ' set but file instances are not identical'
        )

  # Check that the merged FileDescriptorSet is valid.
  try:
    pool = create_descriptor_pool(result_set)
    for file in result_set.file:
      # FindFileByName forces lazy resolution of the file and its dependencies,
      # which validates that all dependencies are present.
      pool.FindFileByName(file.name)
  except Exception as e:
    raise ValueError(
        'failed to generate a valid merged FileDescriptorSet (sets were likely'
        ' built at different times).'
    ) from e

  return result_set


def _append_file_descriptor_set_to_pool(
    pool: descriptor_pool.DescriptorPool,
    file_set: descriptor_pb2.FileDescriptorSet,
) -> None:
  """Appends a given file descriptor set to a file descriptor pool.

  Runs through all file descriptors and appends them one-by-one to the pool.

  Args:
    pool: Protobuf descriptor pool to append descriptors to
    file_set: proto containing the set of descriptors to add.
  """
  file_by_name = {file_proto.name: file_proto for file_proto in file_set.file}

  # N.B. GetMessages() file protos must be added in topo ordering to satisfy the
  # cpp impl of python Protobuf. If this is not done, we get strange errors
  # like "Message.field: 'FieldType' seems to be defined in 'file.proto',
  # which is not imported by '#.proto'.  To use it here, please add the
  # necessary import"
  def add_file(file_proto: descriptor_pb2.FileDescriptorProto):
    for dependency in file_proto.dependency:
      if dependency in file_by_name:
        # Remove from elements to be visited, in order to cut cycles.
        add_file(file_by_name.pop(dependency))
    pool.Add(file_proto)

  while file_by_name:
    add_file(file_by_name.popitem()[1])


def create_descriptor_pool(
    file_set: descriptor_pb2.FileDescriptorSet,
) -> descriptor_pool.DescriptorPool:
  """Returns a DescriptorPool containing all files in the given file set.

  Args:
    file_set: The file descriptors to seed the descriptor pool with.

  Returns:
    A descriptor pool containing all descriptors from the given file_set.
  """
  pool = descriptor_pool.DescriptorPool()
  _append_file_descriptor_set_to_pool(pool, file_set)
  return pool


def _add_location_with_path(
    path: list[int],
    full_name: str,
    source_code_info: descriptor_pb2.SourceCodeInfo,
    locations: dict[str, descriptor_pb2.SourceCodeInfo.Location],
):
  for location in source_code_info.location:
    if path == location.path and (
        location.leading_detached_comments
        or location.leading_comments
        or location.trailing_comments
    ):
      locations[full_name] = location


def _extract_locations_from_message_type(
    message_type: descriptor_pb2.DescriptorProto,
    message_full_name: str,
    path: list[int],
    source_code_info: descriptor_pb2.SourceCodeInfo,
    locations: dict[str, descriptor_pb2.SourceCodeInfo.Location],
):
  _add_location_with_path(path, message_full_name, source_code_info, locations)

  for i, nested_message_type in enumerate(message_type.nested_type):
    nested_path = path + [
        descriptor_pb2.DescriptorProto.NESTED_TYPE_FIELD_NUMBER,
        i,
    ]
    _extract_locations_from_message_type(
        nested_message_type,
        f'{message_full_name}.{nested_message_type.name}',
        nested_path,
        source_code_info,
        locations,
    )

  for i, enum_type in enumerate(message_type.enum_type):
    enum_path = path + [
        descriptor_pb2.DescriptorProto.ENUM_TYPE_FIELD_NUMBER,
        i,
    ]
    _extract_locations_from_enum_type(
        enum_type,
        f'{message_full_name}.{enum_type.name}',
        enum_path,
        source_code_info,
        locations,
    )

  for i, field in enumerate(message_type.field):
    field_path = path + [descriptor_pb2.DescriptorProto.FIELD_FIELD_NUMBER, i]
    _add_location_with_path(
        field_path,
        f'{message_full_name}.{field.name}',
        source_code_info,
        locations,
    )


def _extract_locations_from_enum_type(
    enum_type: descriptor_pb2.EnumDescriptorProto,
    enum_full_name: str,
    path: list[int],
    source_code_info: descriptor_pb2.SourceCodeInfo,
    locations: dict[str, descriptor_pb2.SourceCodeInfo.Location],
):
  _add_location_with_path(path, enum_full_name, source_code_info, locations)

  for i, value in enumerate(enum_type.value):
    value_path = path + [
        descriptor_pb2.EnumDescriptorProto.VALUE_FIELD_NUMBER,
        i,
    ]
    value_full_name = f'{enum_full_name}.{value.name}'
    _add_location_with_path(
        value_path, value_full_name, source_code_info, locations
    )


def _extract_locations_from_file(
    file: descriptor_pb2.FileDescriptorProto,
    locations: dict[str, descriptor_pb2.SourceCodeInfo.Location],
):
  for i, message_type in enumerate(file.message_type):
    path = [descriptor_pb2.FileDescriptorProto.MESSAGE_TYPE_FIELD_NUMBER, i]
    message_full_name = f'{file.package}.{message_type.name}'
    _extract_locations_from_message_type(
        message_type, message_full_name, path, file.source_code_info, locations
    )

  for i, enum_type in enumerate(file.enum_type):
    path = [descriptor_pb2.FileDescriptorProto.ENUM_TYPE_FIELD_NUMBER, i]
    enum_full_name = f'{file.package}.{enum_type.name}'
    _extract_locations_from_enum_type(
        enum_type, enum_full_name, path, file.source_code_info, locations
    )

  return None


def extract_comment_locations_from_file_descriptor_set(
    fds: descriptor_pb2.FileDescriptorSet,
) -> dict[str, descriptor_pb2.SourceCodeInfo.Location]:
  """Extracts comment locations from a file descriptor set.

  Extracts from the given file descriptor set a mapping from full names to
  corresponding SourceCodeInfo.Location objects. Entries are only generated for
  source code elements which have associated comments and only for the following
  types of source code elements:
    - messages    (example key: "intrinsic_proto.Pose")
    - fields      (example key: "intrinsic_proto.Pose.x")
    - enums       (example key: "intrinsic_proto.UpdateMode")
    - enum values (example key: "intrinsic_proto.UpdateMode.OVERWRITE")

  For example, assume the input is a file descriptor set created from the
  following, single proto file:

  ```
  syntax = "proto3";
  package intrinsic_proto;

  // Number detached comment

  // Number leading comment
  message Number { // Number trailing comment
  }
  ```

  Then the result would be:

  {
    "intrinsic_proto.Number": descriptor_pb2.SourceCodeInfo.Location(
      leading_detached_comments=[' Number detached comment\n'],
      leading_comments=' Number leading comment\n',
      trailing_comments=' Number trailing comment\n',
      path: ...,
      span: ...,
    )
  }
  """
  locations: dict[str, descriptor_pb2.SourceCodeInfo.Location] = {}

  for file in fds.file:
    _extract_locations_from_file(file, locations)

  return locations


def extract_attached_comments_from_file_descriptor_set(
    fds: descriptor_pb2.FileDescriptorSet,
) -> dict[str, str]:
  """Extracts attached comments from a file descriptor set.

  This is a convenience variant of
  extract_comment_locations_from_file_descriptor_set() above.

  Extracts from the given file descriptor set a mapping from full names to a
  concatenation of all leading and trailing comments (a multi-line string).
  Entries are only generated for source code elements which have associated
  comments and only for the following types of source code elements:
    - messages    (example key: "intrinsic_proto.Pose")
    - fields      (example key: "intrinsic_proto.Pose.x")
    - enums       (example key: "intrinsic_proto.UpdateMode")
    - enum values (example key: "intrinsic_proto.UpdateMode.OVERWRITE")

  For example, assume the input is a file descriptor set created from the
  following, single proto file:

  ```
  syntax = "proto3";
  package intrinsic_proto;

  // Number leading comment
  message Number { // Number trailing comment
  }
  ```

  Then the result would be:

  {
    "intrinsic_proto.Number":
      "Number leading comment\nNumber trailing comment\n"
  }

  """

  locations = extract_comment_locations_from_file_descriptor_set(fds)

  comments: dict[str, str] = {}
  for full_name, location in locations.items():
    comment = ''
    if location.leading_comments:
      comment += textwrap.dedent(location.leading_comments)
    if location.trailing_comments:
      comment += textwrap.dedent(location.trailing_comments)
    comments[full_name] = comment

  return comments
